src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServer.cs

```diff
@@ -96,7 +104,8 @@ private static ServiceContext CreateServiceContext(IOptions<KestrelServerOptions
                 SystemClock = systemClock,
                 DateHeaderValueManager = dateHeaderValueManager,
                 ConnectionManager = connectionManager,
-                ServerOptions = serverOptions
+                Resources = new ResourceManager(normalConnections, upgradedConnections),
+                ServerOptions = serverOptions,
```

Comment:
What do you think about moving ResourceManager into a filed of ConnectionManager?

I think you could have ConnectionManager.TryAddConnection which calls Resources.NormalConnections.TryLockOne(),  Log.ConnectionStart, EventSource.ConnectionStart etc... in addition to what it does already assuming TryLockOne succeeds. ConnectionManager.RemoveConnection would do the dual.

This would consolidate/encapsulate the connection tracking/logging logic which should make it a little easier to follow in addition to removing a top-level property on ServiceContext. FrameConnectionManager could just take KestrelServerLimits as a ctor param and initialize the ResourceCounters itself.